src/raft/tracker.nim

```diff
@@ -31,67 +36,222 @@ type
+    # log entries to it, but it is not yet counted towards majorities for voting or commitment purposes
+    tracker.current.add(s)
+    let oldp = oldProgress.find(s)
+    if oldp != -1:
+      tracker.progress.add(oldProgress[oldp])
+    else:
+      let progress = RaftFollowerProgressTrackerRef.new(s, nextIndex, now)
+      tracker.progress.add(progress)
+
+  if config.isJoint:
+    for s in config.previousSet:
+      tracker.previous.add(s)
+      var newp = tracker.progress.find(s)
+      if newp != -1:
+        # It already exists in the current set
+        continue
+      let oldp = oldProgress.find(s)
+      if oldp != -1:
+        tracker.progress.add(oldProgress[oldp])
+      else:
+        tracker.progress.add(RaftFollowerProgressTrackerRef.new(s, nextIndex, now))
+
+func init*(
+    T: type RaftTracker,
+    config: RaftConfig,
+    nextIndex: RaftLogIndex,
+    now: times.DateTime,
+): T =
+  var tracker = T()
+  tracker.setConfig(config, nextIndex, now)
+  tracker
+
+func committed*(tracker: RaftTracker, previousCommitIndex: RaftLogIndex): RaftLogIndex =
+  var current = MatchSeqRef.new(previousCommitIndex)
+  if tracker.previous.len != 0:
+    var previous = MatchSeqRef.new(previousCommitIndex)
+    for progress in tracker.progress:
+      if tracker.current.contains(progress.id):
+        current.add(progress.matchIndex)
+      if tracker.previous.contains(progress.id):
+        previous.add(progress.matchIndex)
+    if not current.committed() or not previous.committed():
+      return previousCommitIndex
+    return min(current.commitIndex(), previous.commitIndex())
+  else:
+    for progress in tracker.progress:
+      if tracker.current.contains(progress.id):
+        current.add(progress.matchIndex)
+    if not current.committed():
+      return previousCommitIndex
+    return current.commitIndex()
+
+func accepted*(fpt: var RaftFollowerProgressTrackerRef, index: RaftLogIndex) =
   fpt.matchIndex = max(fpt.matchIndex, index)
-  fpt.nextIndex = max(fpt.nextIndex, index)
\ No newline at end of file
+  fpt.nextIndex = max(fpt.nextIndex, index + 1)
+
+func `$`*(progress: RaftFollowerProgressTrackerRef): string =
+  fmt"""
+    Progress status
+    id: {progress.id}
+    nextIndex: {progress.nextIndex}
+    matchIndex: {progress.matchIndex}
+    commitIndex: {progress.commitIndex}
+    replayedIndex: {progress.replayedIndex}
+    lastMessageAt: {progress.lastMessageAt.format("YYYY:MM:dd:HH:mm:ss:fff")}
+  """
+
+func `$`*(election: RaftElectionTracker): string =
+  fmt"""
+    Election status
+    all: {election.all}
+    responded: {election.responded}
+    granted: {election.granted}
+  """
+func `$`*(tracker: RaftTracker): string =
+  fmt"""
+    Tracker status  
+    current: {tracker.current}
+    previous: {tracker.previous}
+    progress: {tracker.progress}
+  """
+
+func `$`*(cfg: RaftConfig): string =
+  result = "\nConfig State: \n"
```

Comment:
Using `result.add` will be more efficient that `result = result & ...`

Reply:
changed